### ISSUE TITLE
Annotation Problem *.pyi

### ISSUE THREAD

**Issue by Roffild** [pytorch/pytorch#26673]
## 🚀 Feature

It is difficult to control the validity of function arguments when they are in different files.
Very often the information in *.pyi does not match *.py.
https://github.com/pytorch/pytorch/blob/fcd13549f923342a7474993928d86181b75734d3/torch/utils/data/dataloader.pyi#L35
I suggest using Python 2 Annotations:
```python
def optional_unwrap (self, x, y):
   # type: (Optional [int], Optional [int]) -> int
```
You can expand this format to indicate in which file to place annotations.
After that, all *.pyi can be completely assembled in one script. Information will always be up to date.

cc @ezyang @gchanan @zou3519

**Comment by vincentqb** [comment]
(Python 2 support will be dropped #23795)

**Comment by malmaud** [comment]
I think everyone would be onboard to move to inline type annotations once Python 2 support is dropped. The current preference was to use stubs instead of comment-based annotations.

Was there a specific incorrect stub?

**Comment by vincentqb** [comment]
We'll wait for Python 2 EOL, and move all annotation inline using Python 3 syntax.

**Comment by ngoldbaum** [comment]
> Was there a specific incorrect stub?

It looks like `_DataLoaderIter` is now called `_BaseDataLoaderIter` and needs to be renamed and updated in the stubs (and possibly the various subclasses need to be added as well).

More generically, there should probably be a test that checks that the stubs agree with the actual implementation, at least so long as the mypy type information is in a separate file from the implementation.

I'll go ahead and put in a pull request this afternoon that hopefully fixes the specific issue with the type stubs for `_BaseDataLoaderIter`, this is my first pytorch issue so bear with me :)

**Comment by malmaud** [comment]
Thanks! There are some tests for the most important stubs but they are far
from comprehensive at the moment.

**Comment by ngoldbaum** [comment]
I guess ultimately the issue here is that only one file -- `aten/src/ATen/function_wrapper.py` -- is getting type checked by the pytorch CI at the moment (see `mypy-files.txt` in the root of the repo). Fixing this issue in general therefore means making `mypy` run over the whole codebase, which is #16574, and seems like a much bigger job than the narrow issues with the `_DataLoaderIter` pointed out here. I think after #27105 is merged this can be closed as a dupe of #16574?

**Comment by ngoldbaum** [comment]
The specific issue pointed out here has been fixed, so I'm going to close this in favor of #16574 which tracks broader improvements to type correctness.
